lib/Serialization/Serialization.h

```diff
@@ -120,29 +84,140 @@ class Serializer : public SerializerBase {
   /// identifier and non-identifier strings.
   llvm::DenseMap<Identifier, IdentifierID> IdentifierIDs;
 
-  /// A map from DeclContexts to their serialized IDs.
-  llvm::DenseMap<const DeclContext*, DeclContextID> DeclContextIDs;
+  /// All uniqued strings that need to be serialized (identifiers and
+  /// non-identifiers).
+  std::vector<StringRef> StringsToWrite;
+
+  /// The last assigned IdentifierID for uniqued strings from this module.
+  ///
+  /// Note that special module IDs and IDs of special names must not be valid
+  /// IdentifierIDs, except that 0 will always represent the empty identifier.
+  uint32_t /*IdentifierID*/ LastUniquedStringID =
+      serialization::NUM_SPECIAL_IDS - 1;
+
+  /// Helper for serializing entities in the AST block object graph.
+  ///
+  /// Keeps track of assigning IDs to newly-seen entities, and collecting
+  /// offsets when those entities are ready to be serialized.
+  ///
+  /// \tparam T The AST type that's being serialized
+  /// \tparam ID The ID type for encoding into the bitstream
+  /// \tparam RecordCode_ The code for the offsets record in the Index block for
+  /// referring to these entities.
+  template <typename T, typename ID, unsigned RecordCode_>
+  class Serialized {
```

Comment:
The naming is slightly misleading here (although the doc comment clarifies the intent) -- the class is called "Serialized" even though it keeps track of both things that have already been serialized and things that need to be serialized. I don't have a good alternative in mind though 🙁. Perhaps `SerializationBookkeeper`?

Reply:
Yeah, I don't love the name either. I'll go with yours if I can't think of anything better.

Reply:
One more potential name `SerializationState`. I like Brent's `RecordKeeper` suggestion too.